libyul/optimiser/FunctionSpecializer.cpp

```diff
@@ -0,0 +1,183 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libyul/optimiser/ASTCopier.h>
+#include <libyul/optimiser/ASTWalker.h>
+#include <libyul/optimiser/FunctionSpecializer.h>
+#include <libyul/optimiser/NameCollector.h>
+#include <libyul/optimiser/NameDispenser.h>
+#include <libyul/optimiser/OptimiserStep.h>
+#include <libyul/optimiser/UnusedFunctionsCommon.h>
+
+#include <libyul/AsmData.h>
+#include <libyul/AsmDataForward.h>
+#include <libyul/YulString.h>
+
+#include <libsolutil/CommonData.h>
+
+#include <boost/algorithm/cxx11/any_of.hpp>
+
+#include <variant>
+
+using namespace std;
+using namespace solidity::util;
+using namespace solidity::yul;
+
+namespace
+{
+
+/// Helper class that creates a copy of the function definition, replacing the names of the variable
+/// declaration with a new name.
+class FunctionCopier: public ASTCopier
+{
+public:
+	FunctionCopier(
+		map<YulString, YulString> _translations
+	):
+		m_translations(_translations)
+	{}
+	using ASTCopier::operator();
+	YulString translateIdentifier(YulString _name) override;
+private:
+	/// A mapping between old and new names. We replace the names of variable declaration contained
+	/// in the mapping with its new name.
+	map<YulString, YulString> const m_translations;
+};
+
+YulString FunctionCopier::translateIdentifier(YulString _name)
+{
+	if (m_translations.count(_name))
+		return m_translations.at(_name);
+	return _name;
+}
+
+}
+
+FunctionSpecializer::Arguments FunctionSpecializer::specializableArguments(FunctionCall const& _f)
+{
+	auto heuristic = [&](Expression const& _e) -> optional<Expression>
+	{
+		if (holds_alternative<Literal>(_e))
+			return ASTCopier{}.translate(_e);
+		return nullopt;
+	};
+
+	return applyMap(_f.arguments, heuristic);
+}
+
+void FunctionSpecializer::operator()(FunctionCall& _f)
+{
+	ASTModifier::operator()(_f);
+
+	if (m_dialect.builtin(_f.functionName.name))
+		return;
+
+	Arguments arguments = specializableArguments(_f);
+
+	if (boost::algorithm::any_of(arguments, [](auto& _a) { return _a.has_value(); }))
+	{
+		auto newName = m_nameDispenser.newName(_f.functionName.name);
+		yulAssert(newName != _f.functionName.name, "Name dispenser returned the same name.");
+		m_oldToNewMap[_f.functionName.name].emplace_back(make_pair(newName, arguments));
```

Comment:
Does it make sense to search if the specialization is already present? You can use SyntacticallyEqual to compare.